[WIP] rewrite of scraper using new XML feed

gooch opened the pull request:
The City of Busselton have implemented a clean XML feed of DAs that
simplifies the scraping a great deal. It should also reduce any errors,
particularly around DAs that cover more than one lot.

gooch committed:
store the lot number, not used by planningalerts but still useful

gooch committed:
deeper scraping

A recent DA was over two lots, listing only the lot numbers and not the
HSE in the listing page. This requires deeper scraping, fetching the
info page for the DA and looking through the very non-semantic text.

gooch committed:
remove colour from stderr

gooch committed:
rewrite of scraper

The City of Busselton have implemented a clean XML feed of DAs that
simplifies the scraping a great deal. It should also reduce any errors,
particularly around DAs that cover more than one lot.

gooch committed:
rewrite of scraper

The City of Busselton have implemented a clean XML feed of DAs that
simplifies the scraping a great deal. It should also reduce any errors,
particularly around DAs that cover more than one lot.

equivalentideas commented:
Nice one @gooch . I'll have a look at this soon. Were you involved with the appearance of this new feed? ;-)

gooch commented:
Not practically, but causally.

On Thu., 20 Oct. 2016, 11:56 am Luke Bacon, notifications@github.com
wrote:

> Nice one @gooch https://github.com/gooch . I'll have a look at this
> soon. Were you involved with the appearance of this new feed ;-)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/planningalerts-scrapers/city_of_busselton_development_applications/pull/1#issuecomment-255003119,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAMrUraJh_D7Z5hTsyrYsoTrXd8cJ4S3ks5q1uZtgaJpZM4Ka-1h
> .

equivalentideas commented:
@gooch have you got this running on morph.io ? It's super helpful for these PRs if you link to that so I can confirm it works easily :)

gooch commented:
Yes.

https://morph.io/gooch/city_of_busselton_development_applications

equivalentideas commented:
@gooch looks good, but some of the addresses, when it includes a building name at the front, are malformed (missing space before street number) . E.g. "Dunsborough School of Natural Horsemanship88 Genoli Road QUINDALUP WA 6281".

Could you fix those up before we merge please?

equivalentideas commented:
Also, the current info_url value is the generic main page, but should be the specific DA page, e.g. http://www.busselton.wa.gov.au/Developing-Busselton/Public-Consultation/DA160696-Extractive-Industry-Sand-216-Fish-Road-Sabina-River

gooch commented:
I hadn't noticed the malformed name,

There's nothing I can do directly to alter the XML feed, but I believe the officer is following this thread, I'll follow it up with him in any case.

equivalentideas commented:
> There's nothing I can do directly to alter the XML feed, but I believe the officer is following this thread, I'll follow it up with him in any case.

Fantastic :) It would be great if they fix it up and provide the proper links.

Otherwise there might be some hacky way to fix up the address in the script, and scrape the webpage for the proper links to the proposals.

gooch commented:
The missing space sounds like it can be resolved internally, but the info urls will need to be collected by a scraper. That being the case, I'll close the PR, and create a new one once I've implemented the hybrid approach.

equivalentideas commented:
@gooch sounds good. You can just add [WIP] to the title of this one, keep it open and push your changes here. That way we keep all this lovely discussion on the tread :)

equivalentideas commented:
I constantly refer to this great post on PR style :) https://github.com/blog/1943-how-to-write-the-perfect-pull-request

SpatialSpath commented:
I've modded the code so that there will be a space in the address. Unfortunately, I can't write anything that will work out the unique name for the individual page as this is manually done by staff and can be 'very unique'.

equivalentideas commented:
> I've modded the code so that there will be a space in the address. Unfortunately, I can't write anything that will work out the unique name for the individual page as this is manually done by staff and can be 'very unique'.

Sounds like progress to me 🌟  @SpatialSpath Great to see you in here too!

gooch committed:
add Mechanize gem

gooch committed:
adds website uri

gooch committed:
Merge branch 'xml-feed-pr'

gooch committed:
hybrid approach to finding the relevent info..

Since it's not possible to determine at the database level what the
title of the advertising URL will be, this commit takes a hybrid
approach.

The idea is to read the provided XML document, then use the
council_reference value to search the list of advertising documents for
the relevent url.

gooch commented:
Sorry for the messy branch, the original branch had the dev commits squashed, but they came back in a merge.. anyway..

I've done the code for the hybrid approach, as discussed, and it's running at:
https://morph.io/gooch/city_of_busselton_development_applications

@SpatialSpath -- the description field has vanished from the XML feed, I've made the scraper treat this as non-fatal for the time being in order to get it running, but I think it should probably be fatal for field to vanish from the XML document.

SpatialSpath commented:
@gooch description is back in there, have a crack at it now.

gooch commented:
Thanks @SpatialSpath, that's working on my machine, morph.io seems to have a stuck queue, but I assume it will work.

@equivalentideas what's the accepted way of dealing with a fatally bad data (like a missing field in the XML)?
Print on stderr and exit 1?

gooch committed:
white space only

gooch committed:
Minor change to the way errors are handled

Each 'broken' record will print an error, but processing will continue.

equivalentideas commented:
> @equivalentideas what's the accepted way of dealing with a fatally bad data (like a missing field in the XML)?
> Print on stderr and exit 1?

@gooch If an application doesn't have all the required data in the table here https://www.planningalerts.org.au/how_to_write_a_scraper then skip it and use `puts` to print to stdout something like "skipping application missing required data".

By the way, you can print to stdout with `puts`, like here :) https://github.com/planningalerts-scrapers/city_of_ryde/blob/master/scraper.rb#L27

gooch committed:
remove logger, print direct to stdout/stderr

gooch commented:
Is there something you're still waiting on before this can be merged?

equivalentideas commented:
> Is there something you're still waiting on before this can be merged?

Hey @gooch I was just waiting for morph.io to get fixed and the this fell back on the todo list :)

I've checked this and it's looking good. Let's :shipit: !
